File: Medium/C++/Majority_Element_II(229).cpp
class Solution {
public:
    vector<int> majorityElement(vector<int>& nums) {
        vector<int> res;
        int m=0, n=0, countm=0, countn=0;
        for (int &i : nums) {
            if (i==m) ++countm;
            else if (i==n) ++countn;
            else if (countm==0) m=i, countm=1;
            else if (countn==0) n=i, countn=1;
            else --countm, --countn;
        }
        countm = countn = 0;
        for (int &i : nums) {
            if (i==m) ++countm;
            else if (i==n) ++countn;
        }
        if (countm > nums.size()/3) res.push_back(m);
        if (countn > nums.size()/3) res.push_back(n);
        return res;
    }
};

File: Medium/Java/Majority_Element_II(229).java
class Solution {
    public List<Integer> majorityElement(int[] nums) {
        List<Integer> res = new ArrayList<>();
        int m=0, n=0, cm=0, cn=0;
        for (int i : nums) {
            if (i==m) ++cm;
            else if (i==n) ++cn;
            else if (cm==0) { m=i; cm=1; }
            else if (cn==0) { n=i; cn=1; }
            else { --cm; --cn; }
        }
        // check if more than n/3
        cm = cn = 0;
        for (int i : nums) {
            if (i==m) ++cm;
            else if (i==n) ++cn;
        }
        if (cm > nums.length/3) res.add(m);
        if (cn > nums.length/3) res.add(n);
        return res;
    }
}